lib/os/cbprintf_nano.c

```diff
@@ -37,77 +37,72 @@ typedef uint32_t uint_value_type;
 BUILD_ASSERT(sizeof(uint_value_type) <= 8U,
 	     "DIGITS_BUFLEN formula may be incorrect");
 
+#define OUTC(_c) do { \
+	out((int)(_c), ctx); \
+	if (IS_ENABLED(CONFIG_CBPRINTF_LIBC_SUBSTS)) { \
+		++count; \
+	} \
+} while (false)
+
 static void print_digits(cbprintf_cb out, void *ctx, uint_value_type num,
 			 unsigned int base, bool pad_before, char pad_char,
 			 int min_width, size_t *countp)
 {
 	size_t count = 0;
 	char buf[DIGITS_BUFLEN];
-	unsigned int i;
+	int i = 0;
 
-	/* Print it backwards into the end of the buffer, low digits first */
-	for (i = DIGITS_BUFLEN - 1U; num != 0U; i--) {
-		buf[i] = "0123456789abcdef"[num % base];
+	/* Print it backwards, low digits first */
+	do {
+		char c = num % base;
+		if (c >= 10) {
+			c += 'a' - '0' - 10;
+		}
+		buf[i++] = c + '0';
 		num /= base;
-	}
-
-	if (i == DIGITS_BUFLEN - 1U) {
-		buf[i] = '0';
-	} else {
-		i++;
-	}
+	} while (num);
 
-	int pad = MAX(min_width - (int)(DIGITS_BUFLEN - i), 0);
+	int pad = MAX(min_width - i, 0);
 
 	for (/**/; pad > 0 && pad_before; pad--) {
-		out(pad_char, ctx);
-		if (IS_ENABLED(CONFIG_CBPRINTF_LIBC_SUBSTS)) {
-			++count;
-		}
-	}
-	for (/**/; i < DIGITS_BUFLEN; i++) {
-		out(buf[i], ctx);
-		if (IS_ENABLED(CONFIG_CBPRINTF_LIBC_SUBSTS)) {
-			++count;
-		}
+		OUTC(pad_char);
 	}
+	do {
```

Comment:
This loop needs to be:
```
	while (i > 0) {
		OUTC(buf[--i]);
	}
```
otherwise you're starting with a character that was never written, which is at least part of the unit test failures in CI.